notifier/amqp/failover.go

```diff
@@ -0,0 +1,94 @@
+package amqp
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+	samqp "github.com/streadway/amqp"
+)
+
+// FailOver will return the first successful connection made against the provided
+// brokers, or an existing connection if not closed.
+//
+// FailOver is safe for concurrent usage.
+type FailOver struct {
+	Config
+	sync.RWMutex
+	conn *samqp.Connection
+}
+
+func (f *FailOver) Connection(ctx context.Context) (*samqp.Connection, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/amqp/failover").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	err := f.Config.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	f.RLock()
+	if f.conn != nil && !f.conn.IsClosed() {
+		log.Debug().Msg("existing connection exist and is not closed. returning this connection")
+		return f.conn, nil
+	}
+	f.RUnlock()
+
+	for _, uri := range f.URIs {
+		// safe to always call DialTLS per docs:
+		// 'DialTLS will use the provided tls.Config when it encounters an amqps:// scheme and will dial a plain connection when it encounters an amqp:// scheme.'
+		conn, err := samqp.DialTLS(uri, &f.tls)
+		if err != nil {
+			if conn != nil {
+				conn.Close()
+			}
+			log.Info().Str("broker", uri).Msg("failed to connect to AMQP broker. attempting next broker")
+			continue
+		}
+		ch, err := conn.Channel()
```

Comment:
Think so, since every error path calls "conn.Close()" and their docs say:

"After returning from this call, all resources associated with this connection,
including the underlying io, Channels, Notify listeners and Channel consumers
will also be closed."